keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,283 @@
+`CachingObject` and operating on objects of that type later.
+
+Note that given that objects are already opaqued in `watchCache`, we
+also avoid unnecessary serialization when GET/LIST requests are served
+from cache.
+
+One open questions is how do we decide what exactly we want to cache.
+Caching everything stored in `watchCache` seems expensive from memory
+usage point of view. What we suggest, is allow configuring individual
+`registries` with exact versions that we want to cache and propagate
+that information down to `Cacher`. That would give us ability to switch
+it on for most demanding resources (e.g. only Node, Endpoints and Pod
+ifrom core/v1 group and only if empty subresource).
+
+Note that based on a [POC][] (slightly different that above design),
+the gains of implementing it include:
+- eliminating kube-apiserver unresponsiveness in case of write of
+a single huge Endpoints object: [#75294#comment-472728088][]
+- ~7% lower cpu-usage
+- XX% less memory allocations
+
+[POC]: https://github.com/kubernetes/kubernetes/pull/60067
+[#75294#comment-472728088]: https://github.com/kubernetes/kubernetes/issues/75294#issuecomment-472728088
+
+### Risks and Mitigations
+
+Given this proposal doesn't introduce any user visible change, the only risk
+are bugs in implementation. Fortunately, the serialization code is widely
+used by all end-to-end tests, so any bugs should be catched by those or
+unit tests of newly added logic.
+
+## Design Details
+
+### Test Plan
+
+* Unit tests covering all corner cases.
+* Regular e2e tests are passing.
+
+### Graduation Criteria
+
+* All existing e2e tests are passing.
+* Scalability tests confirm gains of that change.
+
+Given it is purely implementation detail, we are not going to expose it
+to users via flag or feature-gate (though the implementation will be done
+behind some kind of hard-coded guard).
+
+### Upgrade / Downgrade Strategy
+
+Not applicable
+
+### Version Skew Strategy
+
+Not applicable
+
+## Implementation History
+
+- 2019-03-27: Created
+
+## Drawbacks
+
+This proposal doesn't improve performance of read-requests being served
+from etcd. See [Alternatives](#alternatives-optional) section for more
+details why it was chosen.
+
+## Alternatives
+
+### LRU cache
+We considered using simple LRU cache to store serialized objects.
+
+Pros:
+- performance gains also for reads served from etcd (though these
+doesn't seem to be huge based on experiments)
+
+Cons:
+- potentially significant point of contention
+- no-control over what is still cached (e.g. for frequently changing
+resources, we still keep them in cache, even if they will never be
+served again)
+
+### Smart objects
+We also considered using `smart objects` - an object that carries the
+serialized format of object from etcd with itself.
+
+Pros:
+- very clear encapsulation
+
+Cons:
+- We need an ability to in-place add fields to serialized object
+(i.e. SelfLink) - very tricky and error-prone
```

Comment:
That depends a lot on how you do it, I think. If we e.g. annotate the fields we're interested in, I bet we can make the proto encoding library emit functions which safely do these things. @jennybuckley and @apelisse have just obtained some relevant experience.